# **Penting**
- Pastikan Anda melakukan Run All sebelum mengirimkan submission untuk memastikan seluruh cell berjalan dengan baik.
- Hapus simbol pagar (#) jika Anda menerapkan kriteria tambahan
- Biarkan simbol pagar (#) jika Anda tidak menerapkan kriteria tambahan

# **1. Import Library**
Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [2]:
#Type your code here
import pandas as pd
import numpy as np
import joblib
from google.colab import files
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report

# **2. Memuat Dataset dari Hasil Clustering**
Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [ ]:
# Gunakan dataset hasil clustering yang memiliki fitur Target
# Silakan gunakan dataset data_clustering jika tidak menerapkan Interpretasi Hasil Clustering [Advanced]
# Silakan gunakan dataset data_clustering_inverse jika menerapkan Interpretasi Hasil Clustering [Advanced]

In [5]:
uploaded = files.upload()

Saving data_clustering.csv to data_clustering.csv


In [6]:
df = pd.read_csv('data_clustering.csv')

In [7]:
# Tampilkan 5 baris pertama dengan function head.
df.head()

TransactionAmount  CustomerAge  TransactionDuration  LoginAttempts  \
0           0.007207     0.838710             0.244828            0.0   
1           0.195940     0.806452             0.451724            0.0   
2           0.065680     0.016129             0.158621            0.0   
3           0.096016     0.129032             0.051724            0.0   
4           0.006874     0.430297             0.648276            0.0   

   AccountBalance  TransactionDate_2023-01-02 16:01:14  \
0        0.336832                                False   
1        0.918055                                False   
2        0.068637                                False   
3        0.569198                                False   
4        0.492591                                False   

   TransactionDate_2023-01-02 16:01:26  TransactionDate_2023-01-02 16:06:29  \
0                                False                                False   
1                                False                                False   
2                                False                                False   
3                                False                                False   
4                                False                                False   

   TransactionDate_2023-01-02 16:07:17  TransactionDate_2023-01-02 16:22:54  \
0                                False                                False   
1                                False                                False   
2                                False                                False   
3                                False                                False   
4                                False                                False   

   ...  PreviousTransactionDate_2024-11-04 08:12:18  \
0  ...                                        False   
1  ...                                        False   
2  ...                                        False   
3  ...                                        False   
4  ...                                        False   

   PreviousTransactionDate_2024-11-04 08:12:19  \
0                                        False   
1                                        False   
2                                        False   
3                                        False   
4                                        False   

   PreviousTransactionDate_2024-11-04 08:12:20  \
0                                        False   
1                                        False   
2                                        False   
3                                        False   
4                                        False   

   PreviousTransactionDate_2024-11-04 08:12:21  \
0                                        False   
1                                        False   
2                                        False   
3                                        False   
4                                        False   

   PreviousTransactionDate_2024-11-04 08:12:22  \
0                                        False   
1                                        False   
2                                        False   
3                                        False   
4                                        False   

   PreviousTransactionDate_2024-11-04 08:12:23  \
0                                        False   
1                                        False   
2                                        False   
3                                        False   
4                                        False   

   CustomerAge_binned_Middle-Aged  CustomerAge_binned_Old  Target  Cluster  
0                           False                    True       2        6  
1                           False                    True       2        6  
2                           False                   False       4        5  
3                           False                   False       4        5  
4                            True           

In [8]:
# Pastikan kolom Target ada di dataset
y = df['Target']
X = df.drop(columns=['Target'])

# **3. Data Splitting**
Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [9]:
# Menggunakan train_test_split() untuk melakukan pembagian dataset.
df_train, df_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)
print(f"Jumlah data latih: {df_train.shape[0]}, Jumlah data uji: {df_test.shape[0]}")


Jumlah data latih: 1921, Jumlah data uji: 481


# **4. Membangun Model Klasifikasi**
Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Menggunakan algoritma klasifikasi yaitu Decision Tree.
2. Latih model menggunakan data yang sudah dipisah.

In [12]:
# Buatlah model klasifikasi menggunakan Decision Tree
model_dt = DecisionTreeClassifier(random_state=42)
model_dt.fit(df_train, y_train)

DecisionTreeClassifier(random_state=42)

In [13]:
# Menyimpan Model
# import joblib
# joblib.dump(model, 'decision_tree_model.h5')
joblib.dump(model_dt, 'decision_tree_model.h5')


['decision_tree_model.h5']

In [36]:
files.download('decision_tree_model.h5')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [14]:
# Evaluasi sederhana
y_pred_dt = model_dt.predict(df_test)
print("Evaluasi Decision Tree:")
print(classification_report(y_test, y_pred_dt, digits=4))

Evaluasi Decision Tree:
              precision    recall  f1-score   support

           0     1.0000    1.0000    1.0000       175
           1     1.0000    1.0000    1.0000        91
           2     1.0000    1.0000    1.0000       107
           3     1.0000    1.0000    1.0000        49
           4     1.0000    1.0000    1.0000        59

    accuracy                         1.0000       481
   macro avg     1.0000    1.0000    1.0000       481
weighted avg     1.0000    1.0000    1.0000       481



### 1. Decision Tree
- **Precision, Recall, F1-Score:** Semua kelas mencapai skor 1.0000  
- **Support:** Distribusi data uji seimbang dengan total 481 sampel  

# **5. Memenuhi Kriteria Skilled dan Advanced dalam Membangun Model Klasifikasi**



**Biarkan kosong jika tidak menerapkan kriteria skilled atau advanced**

In [27]:
# Melatih model menggunakan algoritma klasifikasi selain Decision Tree.

In [23]:
results= {}

In [28]:
# Menampilkan hasil evaluasi akurasi, presisi, recall, dan F1-Score pada seluruh algoritma yang sudah dibuat.

In [24]:
# Random Forest
model_rf = RandomForestClassifier(random_state=42)
model_rf.fit(df_train, y_train)
y_pred_rf = model_rf.predict(df_test)
results['RandomForest'] = {
    'accuracy': accuracy_score(y_test, y_pred_rf),
    'precision': precision_score(y_test, y_pred_rf, average='weighted'),
    'recall': recall_score(y_test, y_pred_rf, average='weighted'),
    'f1': f1_score(y_test, y_pred_rf, average='weighted')
}

In [25]:
# Logistic Regression
model_lr = LogisticRegression(max_iter=1000, random_state=42)
model_lr.fit(df_train, y_train)
y_pred_lr = model_lr.predict(df_test)
results['LogisticRegression'] = {
    'accuracy': accuracy_score(y_test, y_pred_lr),
    'precision': precision_score(y_test, y_pred_lr, average='weighted'),
    'recall': recall_score(y_test, y_pred_lr, average='weighted'),
    'f1': f1_score(y_test, y_pred_lr, average='weighted')
}
joblib.dump(model_lr, 'explore_LogisticRegression_classification.h5')

['explore_LogisticRegression_classification.h5']

In [26]:
# Menyimpan Model Selain Decision Tree
# Model ini bisa lebih dari satu
# import joblib
# joblib.dump(___, 'explore_<Nama Algoritma>_classification.h5')

In [29]:
joblib.dump(model_rf, 'explore_RandomForest_classification.h5')

['explore_RandomForest_classification.h5']

In [37]:
files.download('explore_RandomForest_classification.h5')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [30]:
joblib.dump(model_lr, 'explore_LogisticRegression_classification.h5')

['explore_LogisticRegression_classification.h5']

In [38]:
files.download('explore_LogisticRegression_classification.h5')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [31]:
# Tampilkan hasil perbandingan
print("Hasil Evaluasi Semua Model:")
for name, metrics in results.items():
    print(f"--- {name} ---")
    print(f"Accuracy:  {metrics['accuracy']:.4f}")
    print(f"Precision: {metrics['precision']:.4f}")
    print(f"Recall:    {metrics['recall']:.4f}")
    print(f"F1-Score:  {metrics['f1']:.4f}\n")


Hasil Evaluasi Semua Model:
--- RandomForest ---
Accuracy:  1.0000
Precision: 1.0000
Recall:    1.0000
F1-Score:  1.0000

--- LogisticRegression ---
Accuracy:  1.0000
Precision: 1.0000
Recall:    1.0000
F1-Score:  1.0000



### 1. Random Forest
- **Accuracy:** 1.0000  
- **Precision, Recall, F1-Score (weighted):** 1.0000  
- **Interpretasi:** Dengan ensemble learning, Random Forest juga mampu mencapai performa sempurna pada dataset uji.

### 2. Logistic Regression
- **Accuracy:** 1.0000  
- **Precision, Recall, F1-Score (weighted):** 1.0000  
- **Interpretasi:** Model linier sederhana ini mampu memisahkan kelas-kelas dengan sempurna pada data uji.

Hyperparameter Tuning Model

Pilih salah satu algoritma yang ingin Anda tuning

In [ ]:
# Lakukan Hyperparameter Tuning dan Latih ulang.

In [32]:
param_grid = {
    'max_depth': [None, 5, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

grid_search = GridSearchCV(
    DecisionTreeClassifier(random_state=42),
    param_grid,
    scoring='f1_weighted',
    cv=5,
    n_jobs=-1
)

grid_search.fit(df_train, y_train)
best_dt = grid_search.best_estimator_
print(f"Best Params: {grid_search.best_params_}")

Best Params: {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 2}


In [ ]:
# Menampilkan hasil evaluasi akurasi, presisi, recall, dan F1-Score pada algoritma yang sudah dituning.

In [35]:
y_pred_tuned = tuning_model.predict(df_test)
print("Evaluasi Tuning Decision Tree:")
print(classification_report(y_test, y_pred_tuned, digits=4))

Evaluasi Tuning Decision Tree:
              precision    recall  f1-score   support

           0     1.0000    1.0000    1.0000       175
           1     1.0000    1.0000    1.0000        91
           2     1.0000    1.0000    1.0000       107
           3     1.0000    1.0000    1.0000        49
           4     1.0000    1.0000    1.0000        59

    accuracy                         1.0000       481
   macro avg     1.0000    1.0000    1.0000       481
weighted avg     1.0000    1.0000    1.0000       481



###  Tuning Decision Tree
- **Precision, Recall, F1-Score:** Semua kelas mencapai skor 1.0000  
- **Hyperparameter Terpilih:** Parameter grid terbaik menghasilkan model yang tetap mempertahankan akurasi sempurna.  
- **Interpretasi:** Proses hyperparameter tuning tidak mengubah performa karena model dasarnya sudah sempurna.

In [ ]:
# Menyimpan Model hasil tuning
# import joblib
# joblib.dump(model_dt, 'tuning_classification.h5')

In [34]:
tuning_model = best_dt
joblib.dump(tuning_model, 'tuning_classification.h5')
print("Saved: tuning_classification.h5")

Saved: tuning_classification.h5


In [39]:
files.download('tuning_classification.h5')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Kesimpulan Keseluruhan
Semua algoritma (Decision Tree, Random Forest, Logistic Regression) dan model hasil tuning menunjukkan **performa sempurna** pada data uji (akurasi, presisi, recall, dan F1-Score = 1.0000). Ini menandakan:

1. **Data sangat terpisah** antar kelas sehingga setiap model mampu mempelajari pola dengan sempurna.  
2. **Model sederhana (Decision Tree & Logistic Regression)** sudah cukup efektif—penambahan kompleksitas (Random Forest) atau tuning hyperparameter tidak meningkatkan performa.  
3. **Rekomendasi:**  
   - Perlu diverifikasi lebih lanjut dengan data baru atau cross-validation untuk menghindari overfitting.  
   - Eksplorasi fitur tambahan atau dataset yang lebih menantang agar model dapat diuji pada kondisi nyata.  
